File: Assignment 1/Code/binarySearch.py
import time

def binarySearch (arr, low, high, x):
    if high >= low:
        mid = low + (high - low) // 2
        if arr[mid] == x:
            return mid
        elif arr[mid] > x:
            return binarySearch(arr, low, mid-1, x)
        else:
            return binarySearch(arr, mid + 1, high, x)
    else:
        return -1
    
sizes = [100, 200, 400, 800, 1600, 3200, 6400, 12800, 25600, 51200]
a = [1]*2097152

for i in range(0, 10):
    for j in range(0, sizes[i] + 1):
        a[j] = j
    
    start = time.time()
    for k in range(10000):
        binarySearch(a, 0, j-1, j)
    stop = time.time()
    print("Time for size " + str(j) + " is: " + str((stop - start) * 1000) + "ms")

File: Assignment 1/Code/binarySearch.java
package Code;

class binarySearch {
    int binarySearchFunction(int arr[], int low, int high, int x) {
        if (high >= low) {
            int mid = low + (high - low) / 2;
            if (arr[mid] == x)
                return mid;
            if (arr[mid] > x)
                return binarySearchFunction(arr, low, mid - 1, x);
            return binarySearchFunction(arr, mid + 1, high, x);
        }
        return -1;
    }

    public static void main(String args[]) {
        binarySearch ob = new binarySearch();
        int sizes[] = {100, 200, 400, 800, 1600, 3200, 6400, 12800, 25600, 51200};
        int n = sizes.length;

        int[] a = new int[2097152];
        int i, j, k;

        for (i = 0; i < n; i++) {
            for (j = 0; j < sizes[i]; j++)
                a[j] = j;
            
            long start = System.nanoTime();
            for (k = 0; k < 10000; k++) {
                ob.binarySearchFunction(a,0, j-1, j);
            }
            long end = System.nanoTime();
            long time = end - start;
            System.out.println("Time for size " + j + " is: " + time + "ns");
        }
    }
}

